src/feature_builder.hpp

```diff
@@ -109,44 +201,34 @@ struct feature_builder_visitor
                 }
             }
         }
-        if (valid) finalize(feature_builder);
+        if (valid) finalize(feature_builder, feature);
     }
-
-    void operator()(mapbox::geometry::polygon<coordinate_type> const& poly)
+    void apply_geometry_polygon(vtzero::feature const& feature)
     {
-        auto extent = mapbox::geometry::envelope(poly);
-        if (!boost::geometry::intersects(extent, bbox_)) return;
+        std::vector<detail::polygon_ring<CoordinateType>> rings;
+        vtzero::decode_polygon_geometry(feature.geometry(), detail::polygon_handler<CoordinateType>(rings, dx_, dy_, zoom_factor_));
+
         std::vector<mapbox::geometry::polygon<coordinate_type>> result;
-        boost::geometry::intersection(poly, bbox_, result);
         vtzero::polygon_feature_builder feature_builder{layer_builder_};
-        if (feature_.has_id()) feature_builder.set_id(feature_.id());
+        if (feature.has_id()) feature_builder.set_id(feature.id());
         bool valid = false;
-        for (auto const& p : result)
+
+        mapbox::geometry::multi_polygon<CoordinateType> mp;
+        mp.reserve(rings.size());
+        for (auto&& r : rings)
         {
-            for (auto const& ring : p)
+            if (r.type == vtzero::ring_type::outer)
             {
-                if (ring.size() > 3)
-                {
-                    valid = true;
-                    feature_builder.add_ring(static_cast<unsigned>(ring.size()));
-                    for (auto const& pt : ring)
-                    {
-                        feature_builder.set_point(static_cast<int>(pt.x), static_cast<int>(pt.y));
-                    }
-                }
+                mp.emplace_back();
+                mp.back().push_back(std::move(r.ring));
```

Comment:
What about doing the `bbox` check here? If no intersection with the exterior ring, then return early without encoding any of the geometry.